### ISSUE TITLE
can't exec command with spaces

### ISSUE THREAD

**Issue by shumin1027** [bitfield/script#32]
can't exec command with spaces such as: `bash -c 'docker ps'`

`p := script.Exec("bash -c 'docker ps'")`

**Comment by shumin1027** [comment]
This is how I solve this problem：

```
package shell

import (
	"bytes"
	"github.com/bitfield/script"
	"os/exec"
)

type PipeWrap struct {
	*script.Pipe
}

func (p *PipeWrap) ToPipe(s string) *script.Pipe {
	return p.Pipe
}

func (p *PipeWrap) Exec(s string) *script.Pipe {
	if p == nil || p.Error() != nil {
		return p.Pipe
	}
	q := NewPipeWrap()
	//args := strings.Fields(s)
	cmd := exec.Command("bash", "-c", s)
	cmd.Stdin = p.Reader
	output, err := cmd.CombinedOutput()
	if err != nil {
		q.SetError(err)
	}
	return q.WithReader(bytes.NewReader(output))
}
func NewPipeWrap() *PipeWrap {
	PipeWrap := new(PipeWrap)
	PipeWrap.Pipe = script.NewPipe()
	return PipeWrap
}

func Exec(s string) *script.Pipe {
	return NewPipeWrap().Exec(s)
}
```

**Comment by bitfield** [comment]
Thanks for the report @shumin1027! PR welcome, from anyone who wants to tackle this.

**Comment by posener** [comment]
The right solution for this problem is to change the API of `Exec` to a slice of string, exactly as the `exec.Command` API is, and pass it as is to the `exec.Command` call.

**Comment by bitfield** [comment]
That's actually a little bit of a pain, which is one reason why we wrap `exec.Command` to be able to pass it a single string.

**Comment by posener** [comment]
> That's actually a little bit of a pain, which is one reason why we wrap `exec.Command` to be able to pass it a single string.

I'm not sure I fully understand what is the pain point. Having the input as a single string?

**Comment by bitfield** [comment]
Yes, `exec.Command` takes a string and a slice of strings as the command line. In shell scripts we can simply write this as one string, so `script` replicates this.

**Comment by posener** [comment]
Right, but then you can't have a single argument which is composed of
multiple words...

**Comment by bitfield** [comment]
Unless you quote it, of course, which is what this issue is really about.

**Comment by posener** [comment]
Quoting does not help: The input string is passed to [`strings.Field`](https://github.com/bitfield/script/blob/master/filters.go#L116), which is passed to the `exec.Command` in the next line.

Unfortunately, the strings.Fields does not group into a single field quoted words:

```go
fmt.Println(strings.Join(strings.Fields(`a b "c d" e`), ", "))
// Output: a, b, "c, d", e
```

Notice that in this example, `"c` is one string returned from `strings.Fields` and `d"` is another, while for the `command.Exec` you would want `c d` as a single string.

**Comment by NightMachinary** [comment]
[**@shumin1027**](https://github.com/shumin1027) commented on [Oct 10, 2019, 4:35 AM GMT+3:30](https://github.com/bitfield/script/issues/32#issuecomment-540275384 "2019-10-10T01:05:00Z - Replied by Github Reply Comments"):
> This is how I solve this problem：
> 
> ```
> package shell
> 
> import (
> 	"bytes"
> 	"github.com/bitfield/script"
> 	"os/exec"
> )
> 
> type PipeWrap struct {
> 	*script.Pipe
> }
> 
> func (p *PipeWrap) ToPipe(s string) *script.Pipe {
> 	return p.Pipe
> }
> 
> func (p *PipeWrap) Exec(s string) *script.Pipe {
> 	if p == nil || p.Error() != nil {
> 		return p.Pipe
> 	}
> 	q := NewPipeWrap()
> 	//args := strings.Fields(s)
> 	cmd := exec.Command("bash", "-c", s)
> 	cmd.Stdin = p.Reader
> 	output, err := cmd.CombinedOutput()
> 	if err != nil {
> 		q.SetError(err)
> 	}
> 	return q.WithReader(bytes.NewReader(output))
> }
> func NewPipeWrap() *PipeWrap {
> 	PipeWrap := new(PipeWrap)
> 	PipeWrap.Pipe = script.NewPipe()
> 	return PipeWrap
> }
> 
> func Exec(s string) *script.Pipe {
> 	return NewPipeWrap().Exec(s)
> }
> ```

I'm new to go, but I guess there is no way to make that custom Exec method work in the middle of a pipe? (Because all functions in the library return `script.Pipe`, which we can't add an extension method to.)

**Comment by shumin1027** [comment]
> [**@shumin1027**](https://github.com/shumin1027) commented on [Oct 10, 2019, 4:35 AM GMT+3:30](https://github.com/bitfield/script/issues/32#issuecomment-540275384):
> 
> > This is how I solve this problem：
> > ```
> > package shell
> > 
> > import (
> > 	"bytes"
> > 	"github.com/bitfield/script"
> > 	"os/exec"
> > )
> > 
> > type PipeWrap struct {
> > 	*script.Pipe
> > }
> > 
> > func (p *PipeWrap) ToPipe(s string) *script.Pipe {
> > 	return p.Pipe
> > }
> > 
> > func (p *PipeWrap) Exec(s string) *script.Pipe {
> > 	if p == nil || p.Error() != nil {
> > 		return p.Pipe
> > 	}
> > 	q := NewPipeWrap()
> > 	//args := strings.Fields(s)
> > 	cmd := exec.Command("bash", "-c", s)
> > 	cmd.Stdin = p.Reader
> > 	output, err := cmd.CombinedOutput()
> > 	if err != nil {
> > 		q.SetError(err)
> > 	}
> > 	return q.WithReader(bytes.NewReader(output))
> > }
> > func NewPipeWrap() *PipeWrap {
> > 	PipeWrap := new(PipeWrap)
> > 	PipeWrap.Pipe = script.NewPipe()
> > 	return PipeWrap
> > }
> > 
> > func Exec(s string) *script.Pipe {
> > 	return NewPipeWrap().Exec(s)
> > }
> > ```
> 
> I'm new to go, but I guess there is no way to make that custom Exec method work in the middle of a pipe? (Because all functions in the library return `script.Pipe`, which we can't add an extension method to.)

This issue seems to have been fixed [Fix Exec to handle quotes #42](https://github.com/bitfield/script/pull/42)
example : https://github.com/bitfield/script/blob/master/examples/execute/main.go

**Comment by bitfield** [comment]
That would explain why it's closed...
